kernel/src/dynamic_process_loading.rs

```diff
@@ -0,0 +1,958 @@
+            Err(ErrorCode::NOSUPPORT)
+        }
+    }
+
+    fn load(&self) -> Result<(), ErrorCode> {
+        let process_flash = self.new_process_flash.take().ok_or(ErrorCode::FAIL)?;
+        let remaining_memory = self.app_memory.take().ok_or(ErrorCode::FAIL)?;
+
+        // Get the first eight bytes of flash to check if there is another app.
+        let test_header_slice = match process_flash.get(0..8) {
+            Some(s) => s,
+            None => {
+                // Not enough flash to test for another app. This just means
+                // we are at the end of flash, and there are no more apps to
+                // load. => This case is error in loading app by ota_app, because it means that there is no valid tbf header!
+                return Err(ErrorCode::FAIL);
+            }
+        };
+
+        // Pass the first eight bytes to tbfheader to parse out the length of
+        // the tbf header and app. We then use those values to see if we have
+        // enough flash remaining to parse the remainder of the header.
+        let (version, _header_length, entry_length) =
+            match tock_tbf::parse::parse_tbf_header_lengths(
+                test_header_slice.try_into().or(Err(ErrorCode::FAIL))?,
+            ) {
+                Ok((v, hl, el)) => (v, hl, el),
+                Err(tock_tbf::types::InitialTbfParseError::InvalidHeader(_entry_length)) => {
+                    // If we could not parse the header, then we want to skip over
+                    // this app and look for the next one.
+                    return Err(ErrorCode::FAIL);
+                }
+                Err(tock_tbf::types::InitialTbfParseError::UnableToParse) => {
+                    // Since Tock apps use a linked list, it is very possible the
+                    // header we started to parse is intentionally invalid to signal
+                    // the end of apps. This is ok and just means we have finished
+                    // loading apps. =>
+                    return Err(ErrorCode::FAIL);
+                }
+            };
+
+        // Now we can get a slice which only encompasses the length of flash
+        // described by this tbf header.  We will either parse this as an actual
+        // app, or skip over this region.
+        let entry_flash = process_flash
+            .get(0..entry_length as usize)
+            .ok_or(ErrorCode::FAIL)?;
+
+        let capability = create_capability!(crate::capabilities::ProcessManagementCapability);
+
+        let res = self.load_processes(entry_flash, remaining_memory, &capability);
+        let _ = match res {
+            Ok(()) => Ok(()), // maybe set the remaining memory here if we have to change the process_loading function anyway?
+            Err(_) => Err(ErrorCode::FAIL),
+        };
+
+        // reset the global new app start address and length values
+        self.new_app_start_addr.set(0);
+        self.new_app_length.set(0);
```

Comment:
I think this needs to happen even if an error occurs above.

Reply:
I think I fixed that now.